application/src/main/java/org/thingsboard/server/actors/device/DeviceActorMessageProcessor.java

```diff
@@ -626,39 +667,50 @@ private void processRpcResponseStatus(SessionInfoProto sessionInfo, ToDeviceRpcR
         int requestId = responseMsg.getRequestId();
         log.debug("[{}][{}][{}][{}] Processing RPC command response status: [{}]", deviceId, sessionId, rpcId, requestId, status);
         ToDeviceRpcRequestMetadata md = toDeviceRpcPendingMap.get(requestId);
-
-        if (md != null) {
-            JsonNode response = null;
-            if (status.equals(RpcStatus.DELIVERED)) {
-                if (md.getMsg().getMsg().isOneway()) {
-                    toDeviceRpcPendingMap.remove(requestId);
-                    if (rpcSequential) {
-                        systemContext.getTbCoreDeviceRpcService().processRpcResponseFromDeviceActor(new FromDeviceRpcResponse(rpcId, null, null));
-                    }
-                } else {
-                    md.setDelivered(true);
+        if (md == null) {
+            log.warn("[{}][{}][{}][{}] RPC has already been removed from pending map.", deviceId, sessionId, rpcId, requestId);
+            return;
+        }
+        var toDeviceRpcRequest = md.getMsg().getMsg();
+        boolean persisted = toDeviceRpcRequest.isPersisted();
+        boolean oneWayRpc = toDeviceRpcRequest.isOneway();
+        JsonNode response = null;
+        if (status.equals(RpcStatus.DELIVERED)) {
+            if (oneWayRpc) {
+                toDeviceRpcPendingMap.remove(requestId);
+                if (rpcSequential) {
+                    var fromDeviceRpcResponse = new FromDeviceRpcResponse(rpcId, null, null);
+                    systemContext.getTbCoreDeviceRpcService().processRpcResponseFromDeviceActor(fromDeviceRpcResponse);
                 }
-            } else if (status.equals(RpcStatus.TIMEOUT)) {
-                Integer maxRpcRetries = md.getMsg().getMsg().getRetries();
-                maxRpcRetries = maxRpcRetries == null ? systemContext.getMaxRpcRetries() : Math.min(maxRpcRetries, systemContext.getMaxRpcRetries());
-                if (maxRpcRetries <= md.getRetries()) {
-                    toDeviceRpcPendingMap.remove(requestId);
-                    status = RpcStatus.FAILED;
```

Comment:
@YevhenBondarenko during working on this task, I noticed one issue. Trying to understand if this was by design or not:

Imagine that your RPC isn't persistent. In this case, we need to return a response to the device in case RPC is failed because all retry attempts have been exhausted. However, I didn't notice a logic similar to what I have added below in the code:

```
var fromDeviceRpcResponse = new FromDeviceRpcResponse(rpcId, JacksonUtil.toString(response), null);
systemContext.getTbCoreDeviceRpcService().processRpcResponseFromDeviceActor(fromDeviceRpcResponse);
```

This means that our RPC request will never receive the response and will finally be timeout because of server-side timeout. 

I think that we should either send a response to the rule engine or API in this case or disallow the retry mechanism for RPCs that are not persisted.